cpp/src/parquet/statistics.cc

```diff
@@ -32,84 +32,362 @@ using arrow::MemoryPool;
+  }
+};
 
-template <typename DType>
-void TypedRowGroupStatistics<DType>::SetComparator() {
-  comparator_ =
-      std::static_pointer_cast<CompareDefault<DType> >(Comparator::Make(descr_));
+template <>
+struct CompareHelper<Int32Type, false> {
+  static inline bool Compare(int type_length, int32_t a, int32_t b) {
+    const uint32_t ua = a;
+    const uint32_t ub = b;
+    return ua < ub;
+  }
+};
+
+template <>
+struct CompareHelper<Int64Type, false> {
+  static inline bool Compare(int type_length, int64_t a, int64_t b) {
+    const uint64_t ua = a;
+    const uint64_t ub = b;
+    return ua < ub;
+  }
+};
+
+template <>
+struct CompareHelper<Int96Type, false> {
+  static inline bool Compare(int type_length, const Int96& a, const Int96& b) {
+    if (a.value[2] != b.value[2]) {
+      return (a.value[2] < b.value[2]);
+    } else if (a.value[1] != b.value[1]) {
+      return (a.value[1] < b.value[1]);
+    }
+    return (a.value[0] < b.value[0]);
+  }
+};
+
+template <>
+struct CompareHelper<ByteArrayType, false> {
+  static inline bool Compare(int type_length, const ByteArray& a, const ByteArray& b) {
+    const uint8_t* aptr = reinterpret_cast<const uint8_t*>(a.ptr);
+    const uint8_t* bptr = reinterpret_cast<const uint8_t*>(b.ptr);
+    return std::lexicographical_compare(aptr, aptr + a.len, bptr, bptr + b.len);
+  }
+};
+
+template <>
+struct CompareHelper<FLBAType, false> {
+  static inline bool Compare(int type_length, const FLBA& a, const FLBA& b) {
+    const uint8_t* aptr = reinterpret_cast<const uint8_t*>(a.ptr);
+    const uint8_t* bptr = reinterpret_cast<const uint8_t*>(b.ptr);
+    return std::lexicographical_compare(aptr, aptr + type_length, bptr,
+                                        bptr + type_length);
+  }
+};
+
+template <typename DType, bool is_signed = true>
+class TypedComparatorImpl : public TypedComparator<DType> {
+ public:
+  typedef typename DType::c_type T;
+
+  explicit TypedComparatorImpl(int type_length = -1) : type_length_(type_length) {}
+
+  bool CompareInline(const T& a, const T& b) {
+    return CompareHelper<DType, is_signed>::Compare(type_length_, a, b);
+  }
+
+  bool Compare(const T& a, const T& b) override { return CompareInline(a, b); }
+
+  void GetMinMax(const T* values, int64_t length, T* out_min, T* out_max) override {
+    T min = values[0];
+    T max = values[0];
+    for (int64_t i = 1; i < length; i++) {
+      if (CompareInline(values[i], min)) {
+        min = values[i];
+      } else if (CompareInline(max, values[i])) {
+        max = values[i];
+      }
+    }
+    *out_min = min;
+    *out_max = max;
+  }
+
+  void GetMinMaxSpaced(const T* values, int64_t length, const uint8_t* valid_bits,
+                       int64_t valid_bits_offset, T* out_min, T* out_max) override {
+    ::arrow::internal::BitmapReader valid_bits_reader(valid_bits, valid_bits_offset,
+                                                      length);
+    T min = values[0];
```

Comment:
This is adapted from the original version. If it's wrong here, it's wrong there, and we should open an issue to fix separately

https://github.com/apache/arrow/blob/master/cpp/src/parquet/statistics.cc#L240

Reply:
Will do once this is merged so I'll be able to point the proper line.

Reply:
values can never be null. All nulls are encoded in the definition levels. It is also guaranteed that there is at least one element since there is a check in the writer to create statistics only if there is at least one row.

Reply:
This is in the Spaced code path (for Arrow?), so it's possible there is a logical error somewhere

Reply:
ah! I didn't notice that this is from the Spaced code path. Thanks! It should be handled then.

Reply:
By null, I mean logical Arrow::Array null, e.g. the bit in the validity bitmap is not set, then the corresponding cell in the data array is undefined (or implementation defined).

You should check that there is at least one valid row (if not done already).